File: cs330/Public/ssthreads/putFormula.java
public
class Cell extends Observable implements Observer
{
  
    /*...*/
  public void putFormula(Expression e)
  {
    if (theFormula != null)
    {
      Sequence oldReferences = theFormula.collectReferences();
      for (SearchEnumeration p = oldReferences.front();
           p.hasMoreElements(); ) {
        CellName cn = (CellName)p.nextElement();
        Cell c = theSheet.getCell(cn);
        c.deleteObserver (this);
      }
    }

    theFormula = e;
    if (e != null) 
      {
        Sequence newReferences = e.collectReferences();
        for (SearchEnumeration p = newReferences.front();
             p.hasMoreElements(); ) {
          CellName cn = (CellName)p.nextElement();
          Cell c = theSheet.getCell(cn);
          c.addObserver (this);
        }
      }
    theSheet.cellHasNewFormula (this);  /*1*/
  }
    /*...*/

File: cs330/Public/idioms/putFormula.cpp
void Cell::putFormula(Expression* e)
{
  if (theFormula != 0)  
    {  /*1*/
      CellNameSequence oldReferences = theFormula->collectReferences();
      for (CellNameSequence::Position p = oldReferences.front();
           p != 0; p = oldReferences.getNext(p))
        {
          Cell* c = theSheet.getCell(oldReferences.at(p));
          if (c != 0)
            c->removeObserver (this);
        }
      delete theFormula;
    }
  theFormula = e; /*2*/
  if (e != 0) 
    {     /*3*/
      CellNameSequence newReferences = e->collectReferences();
      for (CellNameSequence::Position p = newReferences.front();
           p != 0; p = newReferences.getNext(p))
        {
          Cell* c = theSheet.getCell(newReferences.at(p));
          if (c != 0)
            c->addObserver (this);
        }
    }
  theSheet.cellHasNewFormula (this); /*4*/
}
